File: server/Kneesocks/Frame.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Kneesocks {
    public class Frame {
        public enum kClosingReason {
            Normal          = 1000,
            GoingAway       = 1001,
            ProtocolError   = 1002,
            FrameTypeError  = 1003,
            DataError       = 1007,
            PolicyError     = 1008,
            FrameTooBig     = 1009,
            RequestError    = 1011
        }

        public enum kOpcode {
            Continuation    = 0x0,
            TextFrame       = 0x1,
            BinaryFrame     = 0x2,
            Close           = 0x8,
            Ping            = 0x9,
            Pong            = 0xA
        };

        public kOpcode Opcode { get; set; } = kOpcode.BinaryFrame;
        public bool IsFinal { get; set; } = true;
        public bool IsMasked { get; set; } = false;
        public byte[] Mask { get; set; } = new byte[] { 0, 0, 0, 0 };
        public byte Reserved { get; set; } = 0;

        private int _HeaderLength = 0;
        public int HeaderLength {
            get {
                if(_HeaderLength != 0)
                    return _HeaderLength;

                int length = 2
                    + (BodyLength >= 0x7E && BodyLength <= 0xFFFF ? 2 : 0)
                    + (BodyLength > 0xFFFF                        ? 4 : 0)
                    + (IsMasked                                   ? 4 : 0);

                return (_HeaderLength = length);
            }
        }

        private int _BodyLength = 0;
        public int BodyLength {
            get {
                return Content == null ? _BodyLength : Content.Length;
            }
        }

        public byte[] Content { get; set; } = null;
        public byte[] MaskedContent {
            get {
                long counter = 0;
                return Content.Select(x => (byte)(x ^ Mask[counter++ % 4])).ToArray();
            }
        }

        public byte[] GetBytes() {
            var headerSize = 2;
            var bodySize = (UInt64)Content.LongLength;
            var headerLengthFirstByte = (byte)Content.Length;
            if(bodySize >= 0x7E && bodySize <= 0xFFFF) {
                headerSize += 2;
                headerLengthFirstByte = 126;
            } else if(bodySize > 0xFFFF) {
                headerSize += 8;
                headerLengthFirstByte = 127;
            }
            var headerLengthSize = (int)headerSize - 1;
            if(IsMasked)
                headerSize += 4;

            var returnValue = new byte[(UInt64)headerSize + bodySize];
            returnValue[0] = (byte)(((byte)Opcode % 0x10) 
                                  | ((Reserved % 8) << 4) 
                                  |  (IsFinal ? 0x80 : 0x0));

            returnValue[1] = (byte)(headerLengthFirstByte 
                                 | (IsMasked ? 0x80 : 0x0));

            if(headerLengthFirstByte >= 0x7E) {
                var upperBound = headerLengthFirstByte == 0x7E ? 2 : 8;
                for(var i = 0; i < upperBound; ++i)
                    returnValue[2 + i] |= (byte)((bodySize >> (8*(upperBound - i))) & 0xFF);
            }

            if(IsMasked)
                Array.Copy(Mask, 0, returnValue, headerSize - 4, 4);

            Array.Copy(IsMasked ? MaskedContent : Content, 0L, returnValue, headerSize, Content.LongLength);
            return returnValue;
        }

        public static int HeaderSizeFromBytes(byte[] raw) {
            if(raw.Length < 2)
                throw new FormatException("Need first two bytes to analyze");

            var lengthByte = raw[1] & 0x7F;
            return 2 
                + ((raw[1] & 0x80) != 0 ? 4: 0)
                +  (lengthByte == 0x7E ? 2 : 0) 
                +  (lengthByte == 0x7F ? 4 : 0);
        }

        public static Frame HeaderFromBytes(byte[] raw) {
            if(raw.Length < 2)
                throw new FormatException("Websocket frame cannot be less than two bytes long");

            var rawOpcode = raw[0] & 0x0F;
            if(!Enum.IsDefined(typeof(kOpcode), rawOpcode))
                throw new ArgumentException("Opcode '0x" + rawOpcode.ToString("X") + "' not understood");

            var returnFrame = new Frame {
                IsFinal  = (raw[0] & 0x80) != 0,
                Opcode   = (kOpcode)rawOpcode,
                IsMasked = (raw[1] & 0x80) != 0,
                Reserved = (byte)((raw[0] & 0x70) >> 4)
            };

            ulong bodyLength = raw[1] & 0x7Ful;
            int headerOffset =
                bodyLength < 0x7E
                    ? 1
                    : (bodyLength == 0x7E ? 3 : 9);

            if(raw.Length < headerOffset + 1)
                throw new FormatException("Websocket frame is smaller than expected header size");

            bodyLength = bodyLength < 0x7E ? 0 : bodyLength;
            for(var i = headerOffset - 1; i > 0; --i)
                bodyLength |= (ulong)raw[2 + i] << (8 * (headerOffset - 1 - i));

            if(bodyLength > Int32.MaxValue)
                throw new FormatException("Frame is too large to interpret");

            returnFrame._BodyLength = (int)bodyLength;

            if(returnFrame.IsMasked)
                Array.Copy(raw, headerOffset + 1, returnFrame.Mask, 0, 4);

            return returnFrame;
        }

        public static Frame FromBytes(byte[] raw) {
            var returnFrame = HeaderFromBytes(raw);

            uint expectedFrameLength = (uint)returnFrame.BodyLength + (uint)returnFrame.HeaderLength;
            if(expectedFrameLength < (uint)raw.Length)
                throw new FormatException("Raw frame length ("+ (uint)raw.Length + ") is less than described size ("+ expectedFrameLength + ")");

            returnFrame.Content = new byte[returnFrame.BodyLength];
            Array.Copy(raw, returnFrame.HeaderLength, returnFrame.Content, 0L, (long)returnFrame.BodyLength);
            if(returnFrame.IsMasked)
                returnFrame.Content = returnFrame.MaskedContent;

            return returnFrame;
        }

        public static Frame Closing(kClosingReason status = kClosingReason.Normal, string reason = "") {
            var data = new byte[2 + reason.ByteLength()];
            data[0] = (byte)(((short)status >> 8) & 0xFF);
            data[1] = (byte)((short)status & 0xFF);
            Array.Copy(reason.GetBytes(), 0, data, 2, reason.ByteLength());

            return new Frame {
                Opcode = kOpcode.Close,
                Content = data
            };
        }

        public static Frame Ping(string data = "") {
            return new Frame {
                Opcode = kOpcode.Ping,
                Content = data.GetBytes()
            };
        }

        public static Frame Pong(string data = "") {
            return new Frame {
                Opcode = kOpcode.Pong,
                Content = data.GetBytes()
            };
        }

        public static Frame Pong(Frame pingFrame) {
            pingFrame.Opcode = kOpcode.Pong;
            return pingFrame;
        }
    }
}

File: server/src/sock/frame.cpp
#include "frame.hpp"

static int calc_header_length(uint64_t body_length) {
    return 
        (body_length < 0x7E
            ? 2
            : (body_length <= 0xFFFF)
                ? 4
                : 10
        );
}

sosc::ws::Frame::Frame
    (const std::string& body, kOpCode opcode, bool final) 
{
    this->body = body;
    this->opcode = opcode;
    this->is_final = final;
}

int sosc::ws::Frame::Parse(const std::string& data, std::string* extra) {
    if(data.length() < 2)
        return FRAME_MORE;
    
    int header_length = Frame::GetHeaderLength(data);
    if(data.length() < header_length)
        return FRAME_MORE;
    
    uint64_t body_length = Frame::GetBodyLength(data);
    if(data.length() < header_length + body_length)
        return FRAME_MORE;
    
    this->is_final = (data[0] & 0x80) != 0;
    if((data[0] & 0x70) != 0)
        return FRAME_ERR;
    if(!IsValidOpCode(data[0] & 0x0F))
        return FRAME_ERR;
    
    this->opcode = static_cast<kOpCode>(data[0] & 0x0F);
    this->body = data.substr(header_length);
    
    if((data[1] & 0x80) != 0) {
        int offset = calc_header_length(body_length);
            
        std::string mask = data.substr(offset, 4);
        for(uint64_t i = 0; i < body_length; ++i)
            this->body[i] ^= mask[i % 4];
    }
    
    uint64_t expected_length = header_length + body_length;
    if(expected_length < data.length() && extra != nullptr)
        *extra = data.substr(expected_length);
    else if(extra != nullptr)
        *extra = "";
    
    return FRAME_OK;
}

int sosc::ws::Frame::GetHeaderLength(const std::string& data) {
    if(data.length() < 2)
        return -1;
    
    uint8_t length_byte = data[1] & 0x7F;
    return 2 +
        ((data[1] & 0x80) != 0 ? 4 : 0) +
        ((length_byte == 0x7E) ? 2 : 0) +
        ((length_byte == 0x7F) ? 8 : 0);
}

uint64_t sosc::ws::Frame::GetBodyLength(const std::string& data) {
    if(data.length() < 2)
        return -1;
    if(data.length() < Frame::GetHeaderLength(data))
        return -1;
    
    uint8_t length_byte = data[1] & 0x7F;
    return 
        (length_byte < 0x7E 
            ? length_byte
            : (length_byte == 0x7E
                ? net::ntohv<uint16_t>(data, 2)
                : net::ntohv<uint64_t>(data, 2)
            )
        );
}

bool sosc::ws::Frame::IsTooLarge(const std::string& data) {
    uint64_t header_length = Frame::GetHeaderLength(data),
             body_length = Frame::GetBodyLength(data);
             
    return (header_length + body_length) <= body_length;
}

void sosc::ws::Frame::PingFrame(Frame* frame) {
    frame->opcode = Ping;
    frame->is_final = true;
    frame->body = net::htonv<uint32_t>(0x600DB00B);
}

void sosc::ws::Frame::PongFrame(Frame* pingFrame) {
    pingFrame->opcode = Pong;
}

void sosc::ws::Frame::ClosingFrame
    (Frame* frame, kClosingReason status, const std::string& reason) 
{
    frame->opcode = Close;
    frame->is_final = true;
    frame->body = net::htonv<uint16_t>(status);
}

std::string* sosc::ws::Frame::ToString(std::string* frame) const {
    uint64_t body_length = this->body.length();
    *frame = std::string(2, 0);
    
    (*frame)[0] = (this->is_final ? 0x80 : 0x00) | this->opcode;
    if(body_length < 0x7E) 
        (*frame)[1] = body_length;
    else if(body_length < 0xFFFF) {
        (*frame)[1] = 0x7E;
        *frame += net::htonv<uint16_t>(body_length);
    } else {
        (*frame)[1] = 0x7F;
        *frame += net::htonv<uint64_t>(body_length);
    }
    
    *frame += this->body;
    return frame;
}

std::string sosc::ws::Frame::ToString() const {
    std::string frame;
    this->ToString(&frame);
    return frame;
}
